From 65d60fe08637d13bfdac508e6e01c00cf4e8e2e1 Mon Sep 17 00:00:00 2001
From: Su Dongming <sudongming1@huawei.com>
Date: Thu, 20 Apr 2023 13:48:20 +0800
Subject: [PATCH 0117/1646] [Huawei] drivers/usb/host: add HISI kunpeng 1710
 USB IP's method of registering in ehci/ohci

Offering: HULK
Hi171x_chip_solution inclusion
category: other
bugzilla: NA

----------------------------------------

Add HISI kunpeng 1710 USB IP's method of registering in ehci/ohci.

Signed-off-by: Su Dongming <sudongming1@huawei.com>
---
 drivers/usb/host/ehci-hcd.c | 20 +++++++++++++++++++-
 drivers/usb/host/ohci-hcd.c | 18 ++++++++++++++++++
 2 files changed, 37 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/host/ehci-hcd.c b/drivers/usb/host/ehci-hcd.c
index 586e2735d6e0..d1455c892f50 100644
--- a/drivers/usb/host/ehci-hcd.c
+++ b/drivers/usb/host/ehci-hcd.c
@@ -1329,6 +1329,11 @@ MODULE_LICENSE ("GPL");
 #define XILINX_OF_PLATFORM_DRIVER	ehci_hcd_xilinx_of_driver
 #endif
 
+#ifdef CONFIG_HISI_KUNPENG1710
+#include "ehci-hi1710.c"
+#define HI1710_OF_PLATFORM_DRIVER	ehci_hcd_hi1710_of_driver
+#endif
+
 #ifdef CONFIG_USB_EHCI_HCD_PMC_MSP
 #include "ehci-pmcmsp.c"
 #define	PLATFORM_DRIVER		ehci_hcd_msp_driver
@@ -1380,15 +1385,25 @@ static int __init ehci_hcd_init(void)
 		goto clean3;
 #endif
 
+#ifdef HI1710_OF_PLATFORM_DRIVER
+	retval = platform_driver_register(&HI1710_OF_PLATFORM_DRIVER);
+	if (retval < 0)
+		goto clean4;
+#endif
+
 #ifdef XILINX_OF_PLATFORM_DRIVER
 	retval = platform_driver_register(&XILINX_OF_PLATFORM_DRIVER);
 	if (retval < 0)
-		goto clean4;
+		goto clean5;
 #endif
 	return retval;
 
 #ifdef XILINX_OF_PLATFORM_DRIVER
 	/* platform_driver_unregister(&XILINX_OF_PLATFORM_DRIVER); */
+clean5:
+#endif
+#ifdef HI1710_OF_PLATFORM_DRIVER
+	platform_driver_unregister(&HI1710_OF_PLATFORM_DRIVER);
 clean4:
 #endif
 #ifdef OF_PLATFORM_DRIVER
@@ -1417,6 +1432,9 @@ static void __exit ehci_hcd_cleanup(void)
 #ifdef XILINX_OF_PLATFORM_DRIVER
 	platform_driver_unregister(&XILINX_OF_PLATFORM_DRIVER);
 #endif
+#ifdef HI1710_OF_PLATFORM_DRIVER
+	platform_driver_unregister(&HI1710_OF_PLATFORM_DRIVER);
+#endif
 #ifdef OF_PLATFORM_DRIVER
 	platform_driver_unregister(&OF_PLATFORM_DRIVER);
 #endif
diff --git a/drivers/usb/host/ohci-hcd.c b/drivers/usb/host/ohci-hcd.c
index 73e13e7c2b46..18e7497dda80 100644
--- a/drivers/usb/host/ohci-hcd.c
+++ b/drivers/usb/host/ohci-hcd.c
@@ -1271,6 +1271,11 @@ MODULE_LICENSE ("GPL");
 #define TMIO_OHCI_DRIVER	ohci_hcd_tmio_driver
 #endif
 
+#ifdef CONFIG_HISI_KUNPENG1710
+#include "ohci-hi1710.c"
+#define HI1710_OF_PLATFORM_DRIVER	ohci_hcd_hi1710_of_driver
+#endif
+
 static int __init ohci_hcd_mod_init(void)
 {
 	int retval = 0;
@@ -1315,9 +1320,19 @@ static int __init ohci_hcd_mod_init(void)
 		goto error_tmio;
 #endif
 
+#ifdef HI1710_OF_PLATFORM_DRIVER
+	retval = platform_driver_register(&HI1710_OF_PLATFORM_DRIVER);
+	if (retval < 0)
+		goto error_hi1710;
+#endif
+
 	return retval;
 
 	/* Error path */
+#ifdef HI1710_OF_PLATFORM_DRIVER
+	platform_driver_unregister(&HI1710_OF_PLATFORM_DRIVER);
+error_hi1710:
+#endif
 #ifdef TMIO_OHCI_DRIVER
 	platform_driver_unregister(&TMIO_OHCI_DRIVER);
  error_tmio:
@@ -1348,6 +1363,9 @@ module_init(ohci_hcd_mod_init);
 
 static void __exit ohci_hcd_mod_exit(void)
 {
+#ifdef HI1710_OF_PLATFORM_DRIVER
+	platform_driver_unregister(&HI1710_OF_PLATFORM_DRIVER);
+#endif
 #ifdef TMIO_OHCI_DRIVER
 	platform_driver_unregister(&TMIO_OHCI_DRIVER);
 #endif
-- 
2.17.1

